apps/system/js/bluetooth_v2.js

```diff
@@ -0,0 +1,430 @@
+      this._requestDisableHandler.bind(this));
+
+    /*
+     * In file transfering case:
+     * since System Message can't be listened in two js files within a app,
+     * so we listen here but dispatch events to bluetooth_transfer.js
+     * while getting bluetooth file transfer start/complete system messages
+     */
+    navigator.mozSetMessageHandler('bluetooth-opp-transfer-start',
+      (transferInfo) => {
+        this._setProfileConnected(this.Profiles.OPP, true);
+        window.dispatchEvent(new CustomEvent('bluetooth-opp-transfer-start',
+          {transferInfo: transferInfo}));
+      }
+    );
+
+    navigator.mozSetMessageHandler('bluetooth-opp-transfer-complete',
+      (transferInfo) => {
+        this._setProfileConnected(this.Profiles.OPP, false);
+        window.dispatchEvent(new CustomEvent('bluetooth-opp-transfer-complete',
+          {transferInfo: transferInfo}));
+      }
+    );
+
+    Service.registerState('isEnabled', this);
+    // LazyLoader.load(['js/bluetooth_icon.js',
+    //                  'js/bluetooth_transfer_icon.js',
+    //                  'js/bluetooth_headphone_icon.js'], function() {
+    //   this.icon = new BluetoothIcon(this);
+    //   this.icon.start();
+    //   this.transferIcon = new BluetoothTransferIcon(this);
+    //   this.transferIcon.start();
+    //   this.headphoneIcon = new BluetoothHeadphoneIcon(this);
+    //   this.headphoneIcon.start();
+    // }.bind(this)).catch(function(err) {
+    //   console.error(err);
+    // });
+  },
+
+  /**
+   * Once adapter is available, listen to bluetooth state and
+   * the connect stat of supported profiles.
+   *
+   * we only support two use cases for bluetooth connection:
+   *   1. connecting with a headset
+   *   2. transfering a file to/from another device
+   * So we need to listen to corresponding events to know we are (aren't)
+   * connected, then summarize to an event and dispatch to StatusBar
+   *
+   * @private
+   */
+  _registerListeners: function bt__registerListeners() {
+    if (this._adapter) {
+      this._btStateChangeHandler =
+        this._btAdapterHandler.bind(this);
+      this._adapter.addEventListener('attributechanged',
+        this._btStateChangeHandler);
+
+      // In headset connected case:
+      this._adapter.addEventListener('hfpstatuschanged', (evt) => {
+        this._setProfileConnected(this.Profiles.HFP, evt.status);
+      });
+
+      this._adapter.addEventListener('a2dpstatuschanged', (evt) => {
+        this._setProfileConnected(this.Profiles.A2DP, evt.status);
+      });
+
+      this._adapter.addEventListener('scostatuschanged', (evt) => {
+        this._setProfileConnected(this.Profiles.SCO, evt.status);
+      });
+    }
+  },
+
+  /**
+   * Remove all EventListeners and adapters when default adapter
+   * is changed.
+   */
+  _cleanupAdapterAndListeners: function bt__cleanupAdapterAndListeners() {
+    this._debug('default adapter is removed');
+    this._dispatchAdapterState(false);
```

Comment:
Dispatching the event in the attribute change handler makes more sense as you dispatch the available event there.